convert-macros/src/lib.rs

```diff
@@ -0,0 +1,100 @@
+use std::iter;
+
+use manyhow::{ensure, manyhow, Result};
+use proc_macro_crate::{crate_name, FoundCrate};
+use quote::{format_ident, quote, ToTokens};
+use syn::{parse_quote, Attribute, DeriveInput, Path};
+
+/// Tries to resolve the path to `extism_convert` dynamically, falling back to feature flags when unsuccessful.
+fn convert_path() -> Path {
+    match crate_name("extism") {
+        Ok(FoundCrate::Name(name)) => {
+            let ident = format_ident!("{name}");
+            parse_quote!(::#ident::convert)
+        }
+        Ok(FoundCrate::Itself) => parse_quote!(crate::core),
+        Err(_) => match crate_name("extism-convert").or_else(|_| crate_name("extism_pdk")) {
+            Ok(FoundCrate::Name(name)) => {
+                let ident = format_ident!("{name}");
+                parse_quote!(::#ident)
+            }
+            Ok(FoundCrate::Itself) => parse_quote!(crate),
+            Err(_) => match () {
+                () if cfg!(feature = "extism-path") => parse_quote!(::extism::convert),
+                () if cfg!(feature = "extism-pdk-path") => parse_quote!(::extism_pdk),
+                _ => parse_quote!(::extism_convert),
+            },
+        },
+    }
+}
+
+fn extract_encoding(attrs: &[Attribute]) -> Result<Path> {
+    let encodings: Vec<_> = attrs
+        .iter()
+        .filter(|attr| attr.path().is_ident("encoding"))
+        .collect();
+    ensure!(!encodings.is_empty(), "encoding needs to be specified"; try = "`#[encoding(ToJson)]`");
+    ensure!(encodings.len() < 2, encodings[1], "only one encoding can be specified"; try = "remove `{}`", encodings[1].to_token_stream());
+
+    Ok(encodings[0].parse_args()?)
+}
+
+#[manyhow]
+#[proc_macro_derive(ToBytes, attributes(encoding))]
+pub fn to_bytes(
+    DeriveInput {
+        attrs,
+        ident,
+        mut generics,
+        ..
+    }: DeriveInput,
+) -> Result {
+    let encoding = extract_encoding(&attrs)?;
+    let convert = convert_path();
+    generics.make_where_clause().predicates.push(
+        parse_quote!(for<'__to_bytes_b> #encoding<&'__to_bytes_b Self>: #convert::ToBytes:<'__to_bytes_b>)
+    );
+    generics.params = iter::once(parse_quote!('__to_bytes_a))
+        .chain(generics.params)
+        .collect();
+    let (impl_generics, type_generics, where_clause) = generics.split_for_impl();
+    Ok(quote! {
+        impl #impl_generics #convert::ToBytes<'__to_bytes_a> for #ident #type_generics #where_clause
+        {
+            type Bytes = ::std::vec::Vec<u8>;
```

Comment:
I tried using the type of the encoding, i.e. `type Bytes = <Json<&Self> as ToBytes>::Bytes;` but was unable to figure out the correct lifetimes.

Reply:
For values encoded with serde, it's likely that `Self::Bytes` is already `Vec<u8>`, so I don't think that's a major issue. The `Raw` encoding might use `&[u8]`, but since this is a higher-level feature I think we can ignore the super high performance case.